File: src/2018/aoc2018-03.ts
export {};
import fs from 'fs';

const input: string[] = fs
  .readFileSync(__filename.replace(/\.[jt]s$/, '.txt'), 'utf-8')
  .trim()
  .split(/\r?\n/g);

const parse = (input: string[]) => {
  return input.map((s) =>
    s
      .match(/^#(\d+) @ (\d+),(\d+): (\d+)x(\d+)$/)
      .slice(1)
      .map((x) => +x)
  );
};

const part1 = (input: string[]) => {
  const lines = parse(input);
  const [maxX, maxY] = lines.reduce(
    ([a, b], [_id, x, y, w, h]) => [Math.max(a, x + w + 1), Math.max(b, y + h + 1)],
    [0, 0]
  );
  const map = Array(maxY)
    .fill(0)
    .map(() => Array(maxX).fill(0));
  for (const [_id, x, y, w, h] of lines)
    for (let j = 0; j < h; j++)
      for (let i = 0; i < w; i++)
        map[y + j][x + i]++;
  let result = 0;
  for (let j = 0; j < maxY; j++) for (let i = 0; i < maxX; i++) result += Number(map[j][i] > 1);
  return result;
};

type N = number;
const between = (xb: N, x: N, xeExcl: N) => xb <= x && x < xeExcl;
const intersects = (xb1: N, xeExcl1: N, xb2: N, xeExcl2: N) => between(xb1, xb2, xeExcl1) || between(xb2, xb1, xeExcl2);
const intersects2D = (x: N, y: N, xe: N, ye: N, x2: N, y2: N, xe2: N, ye2: N) =>
  intersects(x, xe, x2, xe2) && intersects(y, ye, y2, ye2);

const part2 = (input: string[]) => {
  const lines = parse(input);
  const overlap: Record<number, boolean> = {};
  for (let i = 0; i < input.length; i++) {
    const [id, x, y, w, h] = lines[i];
    for (let j = i + 1; j < input.length; j++) {
      const [id2, x2, y2, w2, h2] = lines[j];
      if (intersects2D(x, y, x + w, y + h, x2, y2, x2 + w2, y2 + h2)) overlap[id] = overlap[id2] = true;
    }
  }
  return lines.find(([x]) => !overlap[x])[0];
};

console.time('part 1');
console.log(part1(input));
console.timeEnd('part 1');
console.time('part 2');
console.log(part2(input));
console.timeEnd('part 2');

File: src/2018/aoc2018-03.rs
use std::collections::HashSet;
use std::fmt;
use std::fs::read_to_string;
use std::iter::repeat;
use std::ops::{Add, Index, IndexMut, Range};
use regex::Regex;

use crate::helper::time_it;

mod helper;

#[derive(Debug)]
struct Record {
    id: u32,
    xx: Range<usize>,
    yy: Range<usize>,
}

impl Record {
    fn new(id: u32, x: usize, y: usize, w: usize, h: usize) -> Record { Record { id, xx: x..x + w, yy: y..y + h } }
}

trait Dimension {
    fn width(&self) -> usize;
    fn height(&self) -> usize;
}

impl Dimension for Vec<Record> {
    fn width(&self) -> usize {
        self.iter().map(|r| r.xx.end + 1).max().unwrap()
    }

    fn height(&self) -> usize {
        self.iter().map(|r| r.yy.end + 1).max().unwrap()
    }
}

#[derive(Debug)]
struct Map<T>(Vec<Vec<T>>);

impl<T> Map<T>
    where T: Add<Output=T>, T: Clone {
    // create vector of vector filled with 0
    fn new(width: usize, height: usize, zero: T) -> Map<T> {
        Map(repeat(0).take(height)
            .into_iter()
            .map(|_| repeat(zero.clone())
                .take(width)
                .collect::<Vec<T>>())
            .collect::<Vec<Vec<T>>>())
    }
}

impl<T> fmt::Display for Map<T>
    where T: fmt::Display {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        for row in self {   // would need to loop in &self.0 if no IntoIterator for &Map
            for cell in row {
                write!(f, "{cell} ")?;
            }
            writeln!(f)?;
        }
        Ok(())
    }
}

impl<'a, T> IntoIterator for &'a Map<T> {
    type Item = &'a Vec<T>;
    type IntoIter = core::slice::Iter<'a, Vec<T>>;

    fn into_iter(self) -> Self::IntoIter {
        self.0.iter()
    }
}

impl<T> Index<usize> for Map<T> {
    type Output = Vec<T>;

    fn index(&self, index: usize) -> &Self::Output {
        self.0[index].as_ref()
    }
}

impl<T> IndexMut<usize> for Map<T> {
    fn index_mut(&mut self, index: usize) -> &mut Self::Output {
        self.0[index].as_mut()
    }
}

trait RangeExt {
    fn intersects(&self, other: &Self) -> bool;
    fn intersects2d(&self, y: &Self, x2: &Self, y2: &Self) -> bool;
}

impl<T> RangeExt for Range<T>
    where T: PartialOrd,
{
    fn intersects(&self, other: &Self) -> bool {
        self.contains(&other.start) || other.contains(&self.start)
    }

    fn intersects2d(&self, y: &Self, x2: &Self, y2: &Self) -> bool {
        self.intersects(x2) && y.intersects(y2)
    }
}

fn get_lines() -> Vec<Record> {
    let re = Regex::new(r"^#(\d+) @ (\d+),(\d+): (\d+)x(\d+)$").unwrap();
    read_to_string(file!().replace(".rs", ".txt")).unwrap()
        .split('\n')
        .filter(|s| s.len() > 1)
        .map(|s| {
            let caps = re.captures(s).unwrap();
            let nbs: Vec<usize> = caps.iter()
                .map(|c| c.unwrap().as_str())
                .skip(1)
                .map(|c| c.parse::<usize>().unwrap()).collect();
            Record::new(nbs[0] as u32, nbs[1], nbs[2], nbs[3], nbs[4])
        })
        .collect()
}

fn part1() {
    let records = get_lines();
    let width = records.width();
    let height = records.height();
    let mut map: Map<usize> = Map::new(width, height, 0);
    for rec in records {    // NB: records is moved
        for j in rec.yy {
            for i in rec.xx.clone() {   // NB: no into_iterator on &Range
                map[j][i] += 1;
            }
        }
    }
    let mut result = 0;
    for j in 0..height {
        for i in 0..width {
            result += <bool as Into<usize>>::into(map[j][i] > 1);
        }
    }
    println!("Part 1: {result}");
}

fn part2() {
    let records = get_lines();
    let mut overlap: HashSet<u32> = HashSet::new();
    for i in 0..records.len() {
        let a = &records[i];
        for b in &records[i + 1..] {
            if a.xx.intersects2d(&a.yy, &b.xx, &b.yy) {
                overlap.insert(a.id);
                overlap.insert(b.id);
            }
        }
    }
    let result = records.iter().find_map(|r| if !overlap.contains(&r.id) { Some(r.id) } else { None }).unwrap();
    println!("Part 2: {result}");
}

fn main() {
    time_it(part1);
    time_it(part2);
}
